pkg/networkservice/common/serialize/cancellable_executor.go

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serialize
+
+import (
+	"github.com/edwarnicke/serialize"
+	"github.com/pkg/errors"
+)
+
+// CancellableExecutor is a cancellable serialize.Executor
+type CancellableExecutor struct {
```

Comment:
Note:
1. You can provide a patch to serialize.Executor with adding event on jobs done
2. Use something like
```go
type cancellableExecutor struct {
    executor  serialize.Executor
    counter int32
}
func(e*cancellableExecutor) AsyncExec(f func()) {
  atomic.AddInt32(&e.counter, 1)
  e.executor.AsyncExec(func() {
      f()
      if atomic.AddInt32(&e.counter, -1) == 0 {
            //do cancel logic here     
      }
  })
}
```

Reply:
Actually I don't think that the executor for the connection should be cleaned up before the connection closes. All `serialize/common` complexity comes from the _remove old executor, add new executor for the same ID_ case. Adding new _remove_ points wouldn't make such thing less complicated.
Another point for preserving executor until the connection closes is that it is a way to reject double Close on connection if it happens. We truly know that if there is an executor for the connection, it can be closed. If not, we should stop trying closing. And so it is the same for the after-element Request case - it shouldn't fire if the executor is already closed, because it can't restore the before-chain if it is already closed.
I replaced `CancellableExecutor` with `RequestExecutor`, `CloseExecutor` so now it should be way more better without any `Cancel` functions. https://github.com/networkservicemesh/sdk/blob/c8f5457d2b396b18385d8a3eff5cab2460862566/pkg/networkservice/common/serialize/README.md